### PR TITLE
Fix amount format

### PR DESCRIPTION
**Pull request by nishant-bhorodia**
Hello @colemanw , 

Issue:
if VAT is something like 20.125% it was displayed as  20.125000% on webform.
Also, it displayed errors when submitting, i.e., non_deductible_amount is not a valid amount: 1.20125

Solution:
This PR rounds off the calculated amount to 2 decimal places, hence making sure all percentage amounts for VAT are supported.

Thanks.

### CODE REVIEW

**Comment by colemanw** [comment]
@KarinG do you have any thoughts on this?

**Comment by KarinG** [comment]
Traveling today - but will definitively have a look at this. PS non-deductible amounts are related to income taxes (not sales taxes). I've spotted that confusion before - and have fixed it in CiviCRM core - will fix that in webform civicrm as well.

**Comment by nishant-bhorodia** [comment]
Thanks guys!

**Comment by KarinG** [comment]
Hi @nishant-bhorodia - sorry for the delay - looking at this now.

**Comment by KarinG** [comment]
1. Ok I've set up a test webform:

$10.50 Amount
5% Tax Rate
$10.50 * 1.05 = $11.0775;

image

So yes - perhaps fix the Tax Rate display - so that we would get:

Contribution Amount (includes 20.125% Tax)
Contribution Amount (includes 5% Tax)

the total $ amount here looks good.

2. processing the payment [live transaction into Authorize.net test account] - that's what you're referring to:

image

And that's actually fundamentally wrong - the non-deductible amount should be calculated based on whether or not the financial type is_deductible (or not) [this is in the financial_type configuration in CiviCRM; it applies to income taxes - not sales taxes]

is deductible -> non_deductible_amount = $0
is not deductible -> non_deductible_amount = total_amount

So this code needs to be revised completely. Will draft some code;

File: includes/wf_crm_webform_postprocess.inc [hunk 1]
```diff
@@ -1512,7 +1512,8 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
       // tax integration
       if (!is_null($this->tax_rate)) {
         $item['line_total'] = $item['unit_price'] * (int) $item['qty'];
-        $item['tax_amount'] = ($this->tax_rate / 100) * $item['line_total'];
+        $taxAmount = ($this->tax_rate / 100) * $item['line_total'];
+        $item['tax_amount'] = round($taxAmount, 2);
```
Comment (KarinG) [inline includes/wf_crm_webform_postprocess.inc]: I don't think there is a need to round tax_amount here

File: includes/wf_crm_webform_postprocess.inc [hunk 2]
```diff
@@ -1512,7 +1512,8 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
...
         $this->totalContribution += (1 + $this->tax_rate / 100) * $item['unit_price'] * (int) $item['qty'];
       }
       else {
```
Comment (KarinG) [inline includes/wf_crm_webform_postprocess.inc]: line 1523 - make that 
`return round($this->totalContribution, 2);`

File: includes/wf_crm_webform_postprocess.inc [hunk 3]
```diff
@@ -1881,12 +1882,13 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
...
    */
   private function contributionParams() {
+    $totalContributionAmount = round($this->totalContribution, 2);
     $params = $this->billing_params + $this->data['contribution'][1]['contribution'][1];
     $params['financial_type_id'] = $this->contribution_page['financial_type_id'];
     $params['currency'] = $params['currencyID'] = $this->contribution_page['currency'];
     $params['skipRecentView'] = $params['skipLineItem'] = 1;
     $params['contact_id'] = $this->ent['contact'][1]['id'];
-    $params['total_amount'] = $this->totalContribution;
+    $params['total_amount'] = $totalContributionAmount;
```
Comment (KarinG) [inline includes/wf_crm_webform_postprocess.inc]: round here
`$params['total_amount'] = round($this->totalContribution, 2);`

File: includes/wf_crm_webform_postprocess.inc [hunk 4]
```diff
@@ -1896,8 +1898,9 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
     // tax integration
     // TODO: needs review if this is changed in 4.6
     if (!is_null($this->tax_rate)) {
-      $params['non_deductible_amount']  = $this->totalContribution;
-      $params['tax_amount'] = round(($params['total_amount'] / ($this->tax_rate + 100)) * $this->tax_rate, 2);
+      $params['non_deductible_amount']  = $totalContributionAmount;
+      $taxAmount = ($params['total_amount'] / ($this->tax_rate + 100)) * $this->tax_rate;
+      $params['tax_amount'] = round($taxAmount, 2);
     }
     $params['description'] = t('Webform Payment: @title', array('@title' => $this->node->title));
```
Comment (KarinG) [inline includes/wf_crm_webform_postprocess.inc]: Ok so I think all we need in this loop is:
`$params['tax_amount'] = round(($params['total_amount'] / ($this->tax_rate + 100)) * $this->tax_rate, 2);`

Comment (KarinG) [reply includes/wf_crm_webform_postprocess.inc]: And right after this loop - we need to properly calculate non_deductible amount; I just tested this:
```
// Calculate non-deductible amount for income tax purposes
$financialTypeId = $params['financial_type_id'];
$resultDeduct = wf_civicrm_api('FinancialType', 'get', array('return' => "is_deductible",'id' => $financialTypeId));
if ($resultDeduct['values'][$financialTypeId]['is_deductible'] == 1) {
  // deductible
  $params['non_deductible_amount'] = 0;
} else {
  $params['non_deductible_amount']  = $params['total_amount'];
}

```

**Comment by KarinG** [comment]
My test results using the bits of code/comments I referenced:

Total amount (before rounding was): $11.0775
Tax amount (before rounding was): $0.5275

with Financial Type configuration at Deductible? = Yes

image

with Financial Type configuration at Deductible?  = No

image

**Comment by nishant-bhorodia** [comment]
Hello @KarinG , 
Sorry for the late reply!
Thank you for going through my PR. I have made the suggested changes.
Regards :)

**Comment by KarinG** [comment]
Hi @nishant-bhorodia - great - did you have a chance to test your specific scenarios? If so what were they? I'll run a few tests as well [to complement].

**Comment by KarinG** [comment]
Ok made some final comments and confirmed that this is all that's needed at this point to fix the rounding - to two decimals. If someone uses this w/ a currency that supports more decimals - then we should revisit this. 

```
diff --git a/includes/wf_crm_webform_postprocess.inc b/includes/wf_crm_webform_postprocess.inc
index da520c2..73b7773 100644
--- a/includes/wf_crm_webform_postprocess.inc
+++ b/includes/wf_crm_webform_postprocess.inc
@@ -1830,7 +1830,9 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
 
     $params['contactID'] = $params['contact_id'];
     $params['currency'] = $params['currencyID'] = $this->contribution_page['currency'];
-    $params['total_amount'] = $this->totalContribution;
+    // 78
+    // $params['total_amount'] = $this->totalContribution;
+    $params['total_amount'] = round($this->totalContribution, 2);
     // Some processors want this one way, some want it the other
     $params['amount'] = $params['total_amount'];
 
@@ -1886,7 +1888,9 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
     $params['currency'] = $params['currencyID'] = $this->contribution_page['currency'];
     $params['skipRecentView'] = $params['skipLineItem'] = 1;
     $params['contact_id'] = $this->ent['contact'][1]['id'];
-    $params['total_amount'] = $this->totalContribution;
+    //78
+    //$params['total_amount'] = $this->totalContribution;
+    $params['total_amount'] = round($this->totalContribution, 2);
 
     // Some processors use this for matching and updating the contribution status
     if (!$this->contributionIsPayLater) {
```

File: includes/wf_crm_webform_postprocess.inc [hunk 5]
```diff
@@ -1896,9 +1895,17 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
     // tax integration
     // TODO: needs review if this is changed in 4.6
     if (!is_null($this->tax_rate)) {
```
Comment (KarinG) [inline includes/wf_crm_webform_postprocess.inc:1897]: We pulled in the non_deductible_amount edits in another PR a few days ago - so your Edits below are no longer necessary.

File: includes/wf_crm_webform_postprocess.inc [hunk 6]
```diff
@@ -1519,8 +1519,7 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
         $this->totalContribution += $item['line_total'] = $item['unit_price'] * (int) $item['qty'];
       }
     }
-    $this->totalContribution = round($this->totalContribution, 2);
-    return $this->totalContribution;
+    return round($this->totalContribution, 2);
   }
```
Comment (KarinG) [inline includes/wf_crm_webform_postprocess.inc:1523]: Sorry - my bad - we can leave this as is.

**Comment by nishant-bhorodia** [comment]
Thank you for the review @KarinG :)
Can this be merged or do we still need to test this more ?
Regards!

**Comment by KarinG** [comment]
So trim your PR (and rebase it) and confirm you have tested this on a client site. Then this is good to go!

**Comment by nishant-bhorodia** [comment]
Thanks @KarinG ,
Closing this PR and creating a new one with final changes here #84 
I tested these changes and they work fine in my scenario. I am able to use any percentage of tax/vat after these changes.
